src/scheduler/fifo.c

```diff
@@ -1242,17 +1242,12 @@ run_job(int pbs_sd, resource_resv *rjob, char *execvnode, int throughput, schd_e
 					snprintf(logbuf, MAX_LOG_SIZE, "Job will run for duration=%s", timebuf);
 					schdlog(PBSEVENT_SCHED, PBS_EVENTCLASS_JOB, LOG_NOTICE, rjob->name, logbuf);
 				}
-				if (throughput)
-					rc = pbs_asyrunjob(pbs_sd, rjob->name, execvnode, NULL);
-				else
-					rc = pbs_runjob(pbs_sd, rjob->name, execvnode, NULL);
 			}
-		} else {
-			if (throughput)
-				rc = pbs_asyrunjob(pbs_sd, rjob->name, execvnode, NULL);
-			else
-				rc = pbs_runjob(pbs_sd, rjob->name, execvnode, NULL);
 		}
+		if (rc > 0 && throughput)
+			rc = pbs_asyrunjob(pbs_sd, rjob->name, execvnode, NULL);
+		else
```

Comment:
update_job_attr() can return either a 0 or a -1 when it isn't able to update the job attribute. The if (rc) block at the end will not be hit when update_job_attr() returns a 0, and that's what we will return out of run_job(). So, the rest of the code will think that run_job() was successful and it ran the job even though it never did because pbs_runjob() never gets called if update_job_attr() returns a 0. So I don't think the if (rc) block at the end handles update_job_attr() failing, we just get lucky half the times when the return value is -1. That's why we should be bailing out right after the call to update_job_attr() if it fails.